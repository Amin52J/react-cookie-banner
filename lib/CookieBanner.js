'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _browserCookieLite = require('browser-cookie-lite');

var _styleUtils = require('./styleUtils');

var _styleUtils2 = _interopRequireDefault(_styleUtils);

var CookieBanner = _react2['default'].createClass({
  displayName: 'CookieBanner',

  propTypes: {
    message: _react2['default'].PropTypes.string.isRequired,
    // shortMessage:    React.PropTypes.string,
    onAccept: _react2['default'].PropTypes.func,
    link: _react2['default'].PropTypes.shape({
      msg: _react2['default'].PropTypes.string,
      url: _react2['default'].PropTypes.string.isRequired
    }),
    cookie: _react2['default'].PropTypes.string,
    dismissOnScroll: _react2['default'].PropTypes.bool,
    closeIcon: _react2['default'].PropTypes.string,
    disableStyle: _react2['default'].PropTypes.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      className: '',
      onAccept: function onAccept() {},
      dismissOnScroll: true,
      cookie: 'accepts-cookies'
    };
  },

  getInitialState: function getInitialState() {
    return {
      acceptsCookies: (0, _browserCookieLite.cookie)(this.props.cookie)
    };
  },

  componentDidMount: function componentDidMount() {
    if (!this.state.acceptsCookies && this.props.dismissOnScroll) {
      window.onscroll = this.onScroll;
    }
  },

  onScroll: function onScroll() {
    // tacit agreement buahaha! (evil laugh)
    this.onAccept();
  },

  onAccept: function onAccept() {
    (0, _browserCookieLite.cookie)(this.props.cookie, true);
    this.setState({ acceptsCookies: (0, _browserCookieLite.cookie)(this.props.cookie) });
    this.props.onAccept({ cookie: this.props.cookie });
    if (this.props.dismissOnScroll) {
      window.onscroll = null;
    }
  },

  getStyle: function getStyle(style) {
    if (!this.props.disableStyle) {
      return _styleUtils2['default'].getStyle(style);
    }
  },

  getCloseButton: function getCloseButton() {
    if (this.props.closeIcon) {
      return _react2['default'].createElement('i', { className: this.props.closeIcon, onClick: this.onAccept, style: this.getStyle('icon') });
    }
    return _react2['default'].createElement(
      'div',
      { className: 'button-close', onClick: this.onAccept, style: this.getStyle('button') },
      'Got it'
    );
  },

  getLink: function getLink() {
    if (this.props.link) {
      return _react2['default'].createElement(
        'a',
        {
          href: this.props.link.url,
          className: 'cookie-link',
          style: this.getStyle('link') },
        this.props.link.msg || 'Learn more'
      );
    }
  },

  getBanner: function getBanner() {
    return _react2['default'].createElement(
      'div',
      { className: this.props.className + ' react-cookie-banner', style: this.getStyle('banner') },
      _react2['default'].createElement(
        'span',
        { className: 'cookie-message', style: this.getStyle('message') },
        this.props.message,
        this.getLink()
      ),
      this.getCloseButton()
    );
  },

  render: function render() {
    return this.state.acceptsCookies ? null : this.getBanner();
  }

});

exports['default'] = CookieBanner;
module.exports = exports['default'];