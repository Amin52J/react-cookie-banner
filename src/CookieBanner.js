import React from 'react';
import {cookie} from 'browser-cookie-lite';
import styleUtils from './styleUtils';

const CookieBanner = React.createClass({

  propTypes: {
    message:         React.PropTypes.string.isRequired,
    // shortMessage:    React.PropTypes.string,
    onAccept:        React.PropTypes.func,
    // onDeny:          React.PropTypes.func,
    learnMore:       React.PropTypes.string,
    dismissOnScroll: React.PropTypes.bool,
    closeIcon:       React.PropTypes.string,
    disableStyle:    React.PropTypes.bool
  },

  getDefaultProps() {
    return {
      className: '',
      onAccept: () => {},
      dismissOnScroll: true
    };
  },

  getInitialState() {
    return {
      acceptsCookies: cookie('accepts-cookies')
    };
  },

  componentDidMount() {
    if (!this.state.acceptsCookies && this.props.dismissOnScroll) {
      window.onscroll = this.onScroll;
    }
  },

  onScroll() {
    // tacit agreement buahaha! (evil laugh)
    this.onAccept();
  },

  onAccept() {
    console.log('accepted');
    cookie('accepts-cookies', true);
    this.setState({acceptsCookies: cookie('accepts-cookies')});
    this.props.onAccept({'accepts-cookies': true});
    if (this.props.dismissOnScroll) {
      window.onscroll = null;
    }
  },

  getStyle(style) {
    if (!this.props.disableStyle) {
      return styleUtils.getStyle(style);
    }
  },

  getCloseButton() {
    if (this.props.closeIcon) {
      return <i className={this.props.closeIcon} onClick={this.onAccept} style={this.getStyle('icon')}/>;
    }
    return <div className='button-close' onClick={this.onAccept} style={this.getStyle('button')}>Got it</div>;
  },

  getLearnMore() {
    if (this.props.learnMore) {
      return <a href={this.props.learnMore} style={this.getStyle('learnMore')}>Learn more</a>;
    }
  },

  getBanner() {
    return (
      <div className={this.props.className + ' react-cookie-banner'} style={this.getStyle('banner')}>
        <span className='cookie-message' style={this.getStyle('message')}>
          {this.props.message}
          {this.getLearnMore()}
        </span>
        {this.getCloseButton()}
      </div>
    );
  },

  render() {
    return this.state.acceptsCookies ? null : this.getBanner();
  }

});

export default CookieBanner;
