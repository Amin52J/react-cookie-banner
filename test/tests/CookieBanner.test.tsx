import * as React from 'react';
import { shallow, mount } from 'enzyme';

import CookieBanner from '../../src';

function resetCookies() {
  const cookies = document.cookie.split(';');

  cookies.forEach(cookie => {
    const eqPos = cookie.indexOf('=');
    const name = eqPos > -1 ? cookie.substr(0, eqPos) : cookie;
    document.cookie = `${name}=;expires=Thu, 01 Jan 1970 00:00:00 GMT`;
  });
};

beforeEach(resetCookies);

describe('secondsSinceExpiration', () => {
  const { getSecondsSinceExpiration } = new CookieBanner({ onAccept: () => {} });

  it('should return "cookieExpiration" if it is an integer', () => {
    expect(getSecondsSinceExpiration(12345)).toBe(12345);
  });

  it('should transform "years", "days" and "hours" into seconds', () => {
    expect(getSecondsSinceExpiration({ years: 1, days: 10, hours: 5 })).toBe(32418000);
  });

  it('should handle missing "years", "days" or "hours"', () => {
    expect(getSecondsSinceExpiration({ days: 10 })).toBe(864000);
    expect(getSecondsSinceExpiration({})).toBe(0);
  });

});

describe('CookieBanner', () => {

  it('should be displayed if no cookies are set', () => {
    const component = shallow(
      <CookieBanner message='cookie message' onAccept={() => {}} />
    );
    expect(component).toMatchSnapshot();
  });

  it('should hide on click', () => {
    const component = mount(
      <CookieBanner message='cookie message' onAccept={() => {}} />
    );

    expect(component.find('.react-cookie-banner')).toHaveLength(1);
    component.find('.button-close').simulate('click');
    expect(component.find('.react-cookie-banner')).toHaveLength(0);
  });

  it('should hide on click when dismissOnScroll is false', () => {
    const component = mount(
      <CookieBanner message='cookie message' onAccept={() => {}} dismissOnScroll={false} />
    );

    expect(component.find('.react-cookie-banner')).toHaveLength(1);
    component.find('.button-close').simulate('click');
    expect(component.find('.react-cookie-banner')).toHaveLength(0);
  });

  it('should be displayed with correct message', () => {
    const component = mount(
      <CookieBanner message='cookie message' onAccept={() => {}} />
    );

    expect(component.find('.cookie-message').text()).toBe('cookie message');
  });

  it('should be replaced with custom child component', () => {
    const MyComponent = () => (
      <div className='my-component' />
    );

    const component = mount(
      <CookieBanner message='cookie message' onAccept={() => {}}>
        <MyComponent />
      </CookieBanner>
    );

    expect(component.find('.my-component')).toHaveLength(1);
  });

  it('should be replaced with custom child component using function', () => {
    const MyOtherComponent = ({ onAccept }) => (
      <div className='my-other-component' onClick={onAccept} />
    );

    const customTrigger = onAccept => <MyOtherComponent onAccept={onAccept} />;

    const component = mount(
      <CookieBanner onAccept={() => {}}>
        {customTrigger}
      </CookieBanner>
    );

    expect(component.find('.my-other-component')).toHaveLength(1);
  });

});
