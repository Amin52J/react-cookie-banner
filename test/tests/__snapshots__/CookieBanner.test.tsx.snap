// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieBanner should be displayed if no cookies are set 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CookieBanner
    buttonMessage="Got it"
    cookie="accepts-cookies"
    cookieExpiration={
      Object {
        "years": 1,
      }
    }
    cookies={
      Cookies {
        "HAS_DOCUMENT_COOKIE": true,
        "cookies": Object {},
        "hooks": undefined,
      }
    }
    dismissOnScroll={true}
    dismissOnScrollThreshold={0}
    message="cookie message"
    onAccept={[Function]}
    styles={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buttonMessage": "Got it",
      "className": undefined,
      "closeIcon": undefined,
      "disableStyle": undefined,
      "link": undefined,
      "message": "cookie message",
      "onAccept": [Function],
      "styles": Object {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buttonMessage": "Got it",
        "className": undefined,
        "closeIcon": undefined,
        "disableStyle": undefined,
        "link": undefined,
        "message": "cookie message",
        "onAccept": [Function],
        "styles": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`build build script generates every needed file 1`] = `
Array [
  "BannerContent.d.ts",
  "BannerContent.js",
  "CookieBanner.d.ts",
  "CookieBanner.js",
  "CookieBannerUniversal.d.ts",
  "CookieBannerUniversal.js",
  "index.d.ts",
  "index.js",
  "styleUtils.d.ts",
  "styleUtils.js",
]
`;
